Arm/Memory/Separate.lean

```diff
@@ -115,6 +115,220 @@ example : mem_subset (BitVec.ofNat 64 (2^64 - 1)) 0#64 0#64 (BitVec.ofNat 64 (2^
+  simp [mem_legal] at ha hb
+  rw [BitVec.le_def] at ha hb
+  rw [BitVec.le_def]
+  rw [BitVec.lt_def, BitVec.gt_def] at h
+  rcases h with h | h
+  · sorry
+  · sorry
+  /-
+   by_cases h₅ : a2.toNat < b1.toNat
+   · simp [h₅]
+   · by_cases h₆ : a1.toNat > b2.toNat
+     · simp [h₆]
+     · exfalso
+       rw [BitVec.toNat_sub_eq_toNat_sub_toNat_of_le ha] at h₁ h₂
+       rw [BitVec.toNat_sub_eq_toNat_sub_toNat_of_le hb] at h₃ h₄
+       have h₅' : b1.toNat ≤ a2.toNat := by omega
+       have h₆' : a1.toNat ≤ b2.toNat := by omega
+       rw [BitVec.toNat_sub_eq_toNat_sub_toNat_of_le h₅'] at h₄
+       rw [BitVec.toNat_sub_eq_toNat_sub_toNat_of_le h₆'] at h₂
+       omega
+  -/
+
+/--
+If we express a memory region as `[a..(a+n)]` for `(n : Nat)`,
+and this memory region is legal, then we could not have had any wraparound.
+Thus, it must be the case that (a + n).toNat = a.toNat + n
+-/
+theorem add_lt_of_mem_legal_of_lt
+    (h : mem_legal a (a + n)) :
+    a.toNat + n.toNat < 2^64 := by
+  simp [mem_legal] at h
+  simp [BitVec.le_def, BitVec.toNat_add] at h
+  by_cases hadd : a.toNat + n.toNat < 2^64
+  · assumption
+  · exfalso
+    have ha : a.toNat < 2^64 := a.isLt;
+    have h₂ : a.toNat + n.toNat - 2^64 < 2^64 := by omega
+    have hmod : (a.toNat + n.toNat) % 18446744073709551616 = a.toNat + n.toNat - 2^64 := by
+      rw [Nat.mod_eq_sub_mod]
+      rw [Nat.mod_eq_of_lt h₂]
+      omega
+    rw [hmod] at h
+    omega
+
+/--
+If we express a memory region as `[a..(a+n)]` for `(n : Nat)`,
+and this memory region is legal, then we could not have had any wraparound.
+Thus, it must be the case that (a + n).toNat = a.toNat + n
+-/
+theorem toNat_add_distrib_of_mem_legal_of_lt
+    (h : mem_legal a (a + n)) :
+    (a + n).toNat = a.toNat + n.toNat := by
+  simp [BitVec.add_def]
+  rw [Nat.mod_eq_of_lt]
+  apply add_lt_of_mem_legal_of_lt h
+
 end Separate
 
+
+/-# 
+
+### New Memory Model
+
+The new memory model is different from the old one in two ways:
+
+1. It uses (base pointer, length) to keep track of memory regions instead of closed inntervals of [pointer 1, pointer 2].
+2. To faciliatate the new representation, it bakes in the assumption that the memory region is legal
+   (i.e. no wraparound).
+3. More softly, it tries to keep reasoning in terms of `Nat` rather than `BitVec` in order to allow easier
+   automation via `omega` for proving disjointedness / subset assumptions.
+
+All of the new definitions are named after the old definitions with a prime (') after their name.
+For robustness (and confidence), we plan to prove theorems that establish the equivalence of the old and new memory models.
+-/
+section NewDefitions
+
+/--
+`mem_legal' a n` witnessses that `(a + n)` does not overflow, and thus `[a..a+n)` is a valid range
+of memory. Note that the interval is left closed, right open, and thus `n` is the number of bytes in the memory range.
+-/
+def mem_legal' (a : BitVec 64) (n : Nat) : Prop :=
+  a.toNat + n ≤ 2^64
+
```

Comment:
Presumably you introduce `mem_legal'` because it plays better with your automation, is there a reason not to refactor `mem_legal` to be defined in this way? It's equivalent, right?